domains/client/domain-executor/src/domain_block_processor.rs

```diff
@@ -101,6 +201,123 @@ where
         }
     }
 
+    /// Returns a list of primary blocks waiting to be processed if any.
+    ///
+    /// It's possible to have multiple pending primary blocks that need to be processed in case
+    /// the primary chain re-org occurs.
+    pub(crate) fn pending_imported_primary_blocks(
+        &self,
+        primary_hash: PBlock::Hash,
+        primary_number: NumberFor<PBlock>,
+    ) -> sp_blockchain::Result<Option<PendingPrimaryBlocks<Block, PBlock>>> {
+        if primary_number == One::one() {
+            return Ok(Some(PendingPrimaryBlocks {
+                initial_parent: (self.client.info().genesis_hash, Zero::zero()),
+                primary_imports: vec![HashAndNumber {
+                    hash: primary_hash,
+                    number: primary_number,
+                }],
+            }));
+        }
+
+        let best_hash = self.client.info().best_hash;
+        let best_number = self.client.info().best_number;
+        let best_receipt = crate::aux_schema::load_execution_receipt::<
+            _,
+            Block::Hash,
+            NumberFor<PBlock>,
+            PBlock::Hash,
+        >(&*self.client, best_hash)?
+        .ok_or_else(|| {
+            sp_blockchain::Error::Backend(format!(
+                "Receipt for #{best_number},{best_hash:?} not found"
+            ))
+        })?;
+
+        let primary_from = best_receipt.primary_hash;
+        let primary_to = primary_hash;
+
+        if primary_from == primary_to {
+            return Err(sp_blockchain::Error::Application(Box::from(
+                "Primary block {primary_hash:?} has already been processed.",
+            )));
+        }
+
+        let route =
+            sp_blockchain::tree_route(&*self.primary_chain_client, primary_from, primary_to)?;
+
+        let retracted = route.retracted();
```

Comment:
I don't get what retracted/enacted blocks mean. The documentation here https://paritytech.github.io/substrate/master/sp_blockchain/struct.TreeRoute.html doesn't explain much either. Can you please point me to some info on this?

Reply:
It is just a path across forks over common for both branches block C (in diagram from above link).